horovod/run/run.py

```diff
@@ -0,0 +1,425 @@
+        common_intfs = set(driver.task_addresses_for_tasks(0).keys())
+        for index in range(1, num_hosts):
+            common_intfs.intersection_update(
+                driver.task_addresses_for_tasks(index).keys())
+        if not common_intfs:
+            raise Exception(
+                'Unable to find a set of common task-to-task communication '
+                'interfaces: %s'
+                % [(index, driver.task_addresses_for_tasks(index))
+                   for index in range(num_hosts)])
+        return common_intfs
+    finally:
+        driver.shutdown()
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Horovod Runner')
+
+    parser.add_argument('-v', '--version', action="store_true", dest="version",
+                        help="Show Horovod version.")
+
+    parser.add_argument('-np', '--num-proc', action="store", dest="np",
+                        type=int,
+                        help="Number of processes which should be equal to "
+                             "the total number of available GPUs.")
+
+    parser.add_argument('-p', '--ssh-port', action="store", dest="ssh_port",
+                        type=int,
+                        help="SSH port on all the workers.")
+
+    parser.add_argument('-H', '--host', action="store", dest="host",
+                        help="To specify the list of hosts on which "
+                             "to invoke processes. Takes a comma-delimited "
+                             "list of hosts.")
+
+    parser.add_argument('--disable-cache', action="store_true",
+                        dest="disable_cache",
+                        help="If flag is set, horovod run will not cache the "
+                             "initial checks and execute them every time.")
+    parser.add_argument('--horovod-start-timeout', action="store",
+                        dest="start_timeout",
+                        help="Horovodrun has to perform all the checks and and "
+                             "start before specified timeout.")
+
+    parser.add_argument('--verbose', action="store_true",
+                        dest="verbose",
+                        help="If this flag is set, extra messages will printed.")
+
+    parser.add_argument('command', nargs=argparse.REMAINDER,
+                        help="Command to be executed.")
+
+    return parser.parse_args()
+
+
+def run():
+    args = parse_args()
+
+    if args.version:
+        print(horovod.__version__)
+        exit(0)
+
+    if args.host:
+        host_addresses = [x for x in
+                          [y.split(':')[0] for y in args.host.split(',')]]
+    else:
+        host_addresses = []
+
+    # This cache stores the results of checks performed by horovodrun
+    # during the initialization step. It can be disabled by setting
+    # --disable-cache flag.
+    fn_cache = None
+    if not args.disable_cache:
+        parameters_hash = str(hash(str(args.host) + str(args.np)))
```

Comment:
Using `hash()` in Python3 for this purpose won't work because it's randomized: https://stackoverflow.com/a/17192658.  We should switch to use consistent hashing mechanism, maybe similar to what we use in `host_hash()`.  

We should add a separator between parameters components for cleanliness.  Should we add `.ssh_port` to the hash as well?

Reply:
interesting. Fixed it.